src/wallet/rpcwallet.cpp

```diff
@@ -3085,50 +3087,115 @@ UniValue z_getaddressforaccount(const UniValue& params, bool fHelp)
+            } else {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "pool arguments must be \"transparent\", or \"sapling\"");
             }
         }
     } else {
-        // default is all
-        pools.push_back("transparent");
-        pools.push_back("sapling");
-        pools.push_back("orchard");
+        // Default is the best and second-best shielded pools, and the transparent pool.
+        receivers = {ReceiverType::P2PKH, ReceiverType::Sapling};
     }
+
+    EnsureWalletIsUnlocked();
+
+    // Generate the first UA for this account, using the best and next-best shielded pools
+    // and the transparent pool.
+    auto res = pwalletMain->GenerateUnifiedAddress(account, receivers, j);
+
     UniValue result(UniValue::VOBJ);
-    result.pushKV("account", account);
-    result.pushKV("diversifier_index", params[1].write());
+    result.pushKV("account", (uint64_t)account);
+
+    std::visit(match {
+        [&](std::pair<libzcash::UnifiedAddress, libzcash::diversifier_index_t> addr) {
+            result.pushKV("unifiedaddress", KeyIO(Params()).EncodePaymentAddress(addr.first));
+            result.pushKV("diversifier_index", ArbitraryIntStr(std::vector(addr.second.begin(), addr.second.end())));
+        },
+        [&](AddressGenerationError err) {
+            std::string strErr;
+            switch (err) {
+                case AddressGenerationError::NoSuchAccount:
+                    strErr = tfm::format("Error: account %d has not been generated by z_getnewaccount.", account);
+                    break;
+                case AddressGenerationError::ExistingAddressMismatch:
+                    strErr = tfm::format(
+                        "Error: address at diversifier index %s was already generated with different receiver types.",
+                        params[1].getValStr());
+                    break;
+                case AddressGenerationError::NoAddressForDiversifier:
+                    strErr = tfm::format(
+                        "Error: no address at diversifier index %s was already generated with different receiver types.",
+                        ArbitraryIntStr(std::vector(j.value().begin(), j.value().end())));
+                    break;
+                case AddressGenerationError::InvalidTransparentChildIndex:
+                    strErr = tfm::format(
+                        "Error: diversifier index %s cannot generate an address with a transparent receiver.",
+                        ArbitraryIntStr(std::vector(j.value().begin(), j.value().end())));
+                    break;
+                default:
+                    // By construction, we will not see these errors here:
+                    // - InvalidReceiverTypes
+                    // - WalletEncrypted
+                    //
+                    // If we see these, the user either has generated many addresses, or
+                    // was very unlucky with their mnemonic phrase generation:
```

Comment:
Mnemonic phrase generation checks that it's possible to produce the keys for the default account and the legacy account.

Reply:
Yes, but that would only be relevant if this were the code for generating the default address in `z_getnewaccount`, which you objected to. The generation checks only ensure that a single address can be derived; they don't ensure that more than one can be derived, which is what this RPC method is enabling.